exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/DrillFileSystemCache.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.store.dfs;
+
+import com.google.common.base.Preconditions;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * <h4>Motivation</h4>
+ * <p>
+ *   This cache is intended to work around the bugs in the {@link org.apache.hadoop.fs.FileSystem} static cache (DRILL-5365). Specifically, as of Hadoop 2.7.x the
+ *   {@link org.apache.hadoop.fs.FileSystem} cache has the following bad behavior:
+ * </p>
+ * <ul>
+ *   <li>
+ *     The {@link org.apache.hadoop.conf.Configuration} object is not considered when constructing keys for the {@link org.apache.hadoop.fs.FileSystem} cache of
+ *     {@link org.apache.hadoop.fs.FileSystem} objects.
+ *   </li>
+ *   <li>
+ *      The {@link org.apache.hadoop.fs.FileSystem} cache does not honor the <b>fs.default.name</b> property when constructing keys, only <b>fs.defaultFS</b> is used to construct
+ *      keys in the cache.
+ *   </li>
+ * </ul>
+ *
+ * <h4>Usage</h4>
+ *
+ * <ul>
+ *   <li>
+ *     A prerequisite for usage is that all {@link org.apache.hadoop.conf.Configuration} objects are normalized with
+ *     {@link org.apache.drill.exec.store.dfs.DrillFileSystem#normalize(Configuration)}.
+ *   </li>
+ *   <li>
+ *     This cache should only be used from {@link org.apache.drill.exec.store.dfs.DrillFileSystem}.
+ *   </li>
+ * </ul>
+ *
+ * <h4>TODO</h4>
+ *
+ * <ul>
+ *   <li>
+ *     Drill currently keeps a {@link org.apache.drill.exec.store.dfs.DrillFileSystem} open indefinitely. This will be corrected
+ *     in DRILL-6608. As a result this cache currently has no methods to remove {@link org.apache.hadoop.fs.FileSystem} objects
+ *     after they are created.
+ *   </li>
+ * </ul>
+ */
+class DrillFileSystemCache {
+  private Map<Map<String, String>, FileSystem> cache = new HashMap<>();
```

Comment:
@arina-ielchiieva Thanks for catching this. 

Making the maps concurrent wouldn't be enough, there would still be a race condition in **get** since we are checking and then setting a file system. So we could create two copies of the same file system. I made the get method synchronized instead.

Please take another look.

Reply:
static final?

Reply:
Some how missed this comment last pass. Fixing it now.